chore: use hatch as a build backend for python projects

sfoster1 opened the pull request:
This is a long-time-coming change to switch to using [hatch](https://github.com/pypa/hatch) (or more specifically, [hatchling](https://github.com/pypa/hatch/tree/master/backend)) as a [PEP-517 build backend](https://peps.python.org/pep-0517/) for building our python packages.

## What? Why?
Python packaging has been in a bit of a churn for a while. The way it used to work is you wrote a script, called `setup.py` for convention, that would import utilities from the built in package `distutils` and later its drop in replacement `setuptools` that provided a function called `setup()`; then you called `setup()` with a bunch of descriptive data, and that function would build some kind of python distributable, like an sdist or a wheel (or, back in the day, an egg).

People largely hated this because it meant that you would have to run code to build a package, and that's gross, and also there was no way to figure out how to build a package than writing the code - maybe that `setup.py` had some other dependencies, you'd never know until you tried.

The fix for this was to define an official build system API. Python build systems are now split into backends and frontends. Backends are packages that provide a specific programmatic interface for building packages; frontends are applications that transform user input into the appropriate calls to that interface. These two packages are now commonly configured through the use of a toml file called pyproject.toml with a specific format.

This happened a while ago, and the python ecosystem moved on to supporting it. Specifically, setuptools switched to focusing on that, and doesn't want to deal with setup.py anymore. Features traditionally provided by setuptools to make setup.py work have been getting deprecated or outright broken. Most recently, that was editable-install support. This is even more of a problem because setuptools is commonly included as a transitive dependency without a version bound, so it can just update on you and break stuff.

Switching to use hatchling for building our projects instead of setup.py lets us avoid these breakages by using a more stable build system; it lets us configure our projects using more modern configuration files, that can expand into also configuring our build frontends; and it can all be done without actually changing the way we interact with these projects as devs, because `pip` is a PEP-517 build frontend, and that's what `pipenv` uses to install stuff. So we don't have to switch out pipenv (yet) and everything will Just Keep Working.

## How?
First step is to switch all the python projects to use pyproject.tomls instead of setup.py, requesting hatchling as a build backend. There's two things that don't Just Work with this setup:
1. We get our version from git, which isn't uncommon, but we switch what tag we look at sometimes, which is uncommon. [This hatch plugin](https://github.com/Opentrons/hatch-plugins/tree/main/hatch-vcs-tunable) we wrote lets you override the settings passed to the git version bridge plugin via environment variables
2. We exactly coversion all our dependencies, and we don't want to edit the files to do it. [This hatch plugin](https://github.com/Opentrons/hatch-plugins/tree/main/hatch-dependency-coversion) we wrote automatically overwrites the versions of specified dependencies at build time

We also need to switch [buildroot](https://github.com/Opentrons/buildroot/pull/246) and [openembedded](https://github.com/Opentrons/oe-core/pull/196) to use this.

## What's left?
- system builds working
  - [x] on ot2 https://github.com/Opentrons/buildroot/pull/246
  - [x] on flex https://github.com/Opentrons/oe-core/pull/196
- [x] the other python packages
- verifying that this works all the way through a built robot version
  - [x] on ot2, to boot and run
  - [x] on flex
- verifying that push works
  - [x] on ot2
  - [x] on flex
- [x] verifying that the interned build for offline analysis works
- [ ] verifying that a pypi push works
- [x] checking that everybody is ok with it

sfoster1 committed:
chore: move s-d python up a level

This makes it a lot easier to make an sdist that is as valid as the
source tree for building a wheel out of.

sfoster1 committed:
chore: replace setup.py with pyproject.toml

This will use hatchling, which supports editables in the way we need.

sfoster1 committed:
chore: buildroot integration for new location

sfoster1 committed:
chore: api pipenv linking

sfoster1 committed:
chore: hardware pipenv linking

hardware fixups

sfoster1 committed:
chore: shared-data python tools updates

sfoster1 committed:
chore: root tools and gh updates

root gitignore

sfoster1 committed:
chore: abr-testing updates

sfoster1 committed:
chore: hardware-testing s-d update

sfoster1 committed:
chore: robot-server pipenv update

sfoster1 committed:
chore: test-data-generation pipenv update

sfoster1 committed:
chore: analyses-snapshot-tetsing docker updates

sfoster1 committed:
chore: g-code-testing pipenv and fixes

Also needed to update pydantic, guess it hasn't been locked in a while.
Also asyncio tests were getting autoskipped, and updating them required
fixing some multiprocessing stuff.

sfoster1 committed:
chore: package-testing script fixes

sfoster1 committed:
chore: app-shell analysis install fixes

sfoster1 committed:
chore: hardware pyproject.toml

Update hardware to use pyproject.toml for project configuration and
hatch for building. Uses the special vcs tunable version generation and
the dependency coversioning plugin.

sfoster1 committed:
chore: use hatch in server-utils

sfoster1 committed:
chore: use hatch in update-server

SyntaxColoring commented on abr-testing/Pipfile:
```diff
@@ -11,7 +11,7 @@ types-httplib2 = "*"
 gspread = "==6.0.2"
 hardware-testing = {editable = true, path = "../hardware-testing"}
-opentrons-shared-data = {editable = true, path = "./../shared-data/python"}
+opentrons-shared-data = {editable = true, path = "./../shared-data"}
```
I'm fine with this if we have to do it, but out of curiosity, what is it exactly that is forcing us to move shared-data's Python project config out of the Python subdirectory?

sfoster1 replied:
it's this thing that `build` does that i actually kind of like where it wants `sdist`s to be as similar to the actual, like, source distribution as possible. source distributions are rooted at the config file, so the config file wants to be at or above stuff that's included like the data. and while we can import from above, that makes the wheel build process different if it's built from an sdist vs the actual source

SyntaxColoring commented on api/src/opentrons/_version.py:
Does this get written on-the-fly now by `vcs-tunable`? [Edit: Yes.]

SyntaxColoring commented on hardware/Pipfile:
```diff
@@ -26,7 +27,7 @@ types-mock = "~=5.1.0"
 pytest-asyncio = "~=0.23.0"
 matplotlib = "*"
-opentrons-shared-data = { editable = true, path = "../shared-data/python" }
+build = "~=1.2.0"
```
FYI `~=1.2.0` is equivalent to `>=1.2.0,<1.3`. It seems like we'd want `>=1.2.0,<2` instead?

sfoster1 replied:
eh. honestly i thought about making it unbounded first, i think if we're going to bound it we might as well bound it here

SyntaxColoring replied:
Hmm, why bound it here?

I can understand unbounding it (you trust the `build` maintainers to break things slowly or never), bounding it to 1.x (you trust the `build` maintainers to be semver-compliant), or bounding it to an exact release (you don't trust the `build` maintainers). Bounding it to 1.2.x seems like the worst of all worlds?

This shouldn't block this PR, of course.

sfoster1 replied:
this expresses "I do not want this to fundamentally change but I want to accept weird little bugfixes"

SyntaxColoring commented on update-server/.gitignore:
Other than the `_version.py` entry, can the entries like `.DS_Store` be deduplicated into the top-level `.gitignore`? Or is something in the Python build system looking at these project-specific `.gitignore`s to avoid bundling these in the build?

sfoster1 replied:
yeah sure. the thing i was using just added them as a template

SyntaxColoring commented on shared-data/python/opentrons_shared_data/_version.py:
```diff
@@ -1,20 +1,21 @@
-import sys
-import logging
+# file generated by setuptools-scm
+# don't change, don't track in version control
```
This ought to be deleted and gitignore'd, right?

sfoster1 replied:
definitely

SyntaxColoring reviewed (commented):
Cool!

sfoster1 committed:
chore: use hatch in system-server

sfoster1 committed:
chore: use hatch in usb-bridge

sfoster1 committed:
chore: hatch for test-data-generation

sfoster1 committed:
chore: hatch for performance-metrics

sfoster1 committed:
api pyproject

sfoster1 committed:
chore: hatch for hardware-testing

sfoster1 committed:
sanitize ignores

sfoster1 committed:
more ignores

sfoster1 committed:
buildroot: use hatch and mark deps

sfoster1 committed:
chore(g-code-testing): add some test failure diagnostics

sfoster1 committed:
g-code-testing: more concurrency fixups

sfoster1 committed:
more fixups

sfoster1 committed:
temp: notes for actions

sfoster1 committed:
g-code-testting: use orchestrator, not runner

Using the runner like this means the protocol will never actually run.

sfoster1 committed:
forgot some prints

sfoster1 committed:
fixup some test stuff

- put the results for the tests as artifacts
- why were the diffs presenting the stored data as "new"?????????????

sfoster1 committed:
g-code-testing: remove unsupported unsticking and drops

My old enemy

sfoster1 committed:
pyproject: tighten standin versions

sfoster1 committed:
last gcode update

sfoster1 committed:
fix git describe args for linux

sfoster1 committed:
set wheel versions properly

sfoster1 committed:
Revert "set wheel versions properly"

This reverts commit 7ec21625f866fb0b52598805883d801ddb795d8e.

nevermind this won't work at all

sfoster1 committed:
set wheel version properly, properly

sfoster1 committed:
handle internal-release in buildroot

y3rsh commented on api/pyproject.toml:
```diff
@@ -0,0 +1,84 @@
+description = "The Opentrons API is a simple framework designed to make writing automated biology lab protocols easy."
+
+dependencies = [
+   "opentrons-shared-data==0.0.0",
```
Super nit that the indentation is off on this line.
The 0.0.0 is a placeholder when you have local editable?

y3rsh commented on api/pyproject.toml:
```diff
@@ -0,0 +1,84 @@
+
+[tool.hatch.version.raw-options]
+root = '..'
+git_describe_command = 'git describe --dirty --tags --long --match=v*'
```
I had to add quotes on Mac zsh to do the below.  TIL `git describe`.
```shell
git describe --dirty --tags --long --match="v*"
```
to get 
v8.4.0-0-g241877587a

sfoster1 replied:
yeah, this doesn't get executed through the shell i think so the shell expansion of `*` (which is the issue) doesn't happen

y3rsh commented on api/pyproject.toml:
```diff
@@ -0,0 +1,84 @@
+override-versions-of=['opentrons-shared-data', 'opentrons-hardware', 'opentrons-hardware["FLEX"]']
+
+[tool.hatch.build.hooks.vcs-tunable]
+version-file="src/opentrons/_version.py"
```
We .gitignore this file now but is it still used in the process?

sfoster1 replied:
yeah i'm a little uncomfortable about this but it's nice to have something because otherwise you get setup race conditions occasionally

sfoster1 replied:
it's generated whenever you do an sdist or wheel build

y3rsh commented on api/pyproject.toml:
```diff
@@ -0,0 +1,84 @@
+build-backend = 'hatchling.build'
+
+[tool.hatch.metadata.hooks.dependency-coversion]
+override-versions-of=['opentrons-shared-data', 'opentrons-hardware', 'opentrons-hardware["FLEX"]']
```
Finally following a bit, your docs on https://pypi.org/project/hatch-dependency-coversion/ are great.

sfoster1 replied:
thanks! i decided to go really extra there

sfoster1 committed:
fix various build issues for push

sfoster1 committed:
fix(api): consume smoothie startup text

A word I hoped I'd never have to type again.

When it boots, it gives some intro text. This didn't used to be a
problem but now it was for some reason. Consuming it all, in a very ugly
way, fixes the issue.

sfoster1 committed:
buildroot: better pathing for our makescript

sfoster1 committed:
chore: update push to use pip

This is like 99999 times easier. We don't have to have separate commands
for flex and ot2 anymore, either.

By popping some of the inlined shell code into an actual script it
becomes just a bit more readable.

sfoster1 committed:
chore: g-code-testing lint fixup

sfoster1 committed:
Merge branch 'edge' into hatch-perchance

SyntaxColoring commented on abr-testing/Makefile:
```diff
@@ -53,38 +47,32 @@ clean:
 
 .PHONY: wheel
-wheel: export OPENTRONS_PROJECT=$(project_rs_default)
+wheel: export OPENTRONS_PROJECT=$(PROJECT)
```
I think it's fine for these purposes, but heads up for a Makefile footgun.

If a Makefile has a dependency graph like this:

```mermaid
flowchart TD
    target_1[target_1: export FOO=bar<br>target_1: target_3] --> target_3
    target_2[target_2: target_3] --> target_3
```

...and you run something like `make target_1 target2` (maybe with `-j`), then it's a race condition whether `target_3` runs with or without `FOO=bar`. It has to do with whether `target_3` is first needed for `target_1` or `target_2`; it inherits the environment of whichever dependent asked for it first. Which can be very confusing, as you can imagine.

Something like this ought to avoid the footgun, although it's a little uglier:

```Makefile
wheel:
    rm -rf dist/*.whl
    OPENTRONS_PROJECT=$(PROJECT) \
        HATCH_VCS_TUNABLE_TAG_PATTERN=... \
        HATCH_VCS_TUNABLE_RAW_OPTIONS=... \
        $(python) -m build --wheel .
    $(SHX) rm -rf build
    $(SHX) ls dist
```

Your call, but I would do it even if it doesn't matter in this case because people largely just pattern-match when working on these Makefiles, so it's nice if we can make the patterns inherently safer.

SyntaxColoring commented on usb-bridge/setup.cfg:
🤨 I'm just now realizing that these projects had both setup.py *and* setup.cfg. It's moot now, but was that...okay?

sfoster1 replied:
yeah that's pretty normal for setup.py. you'd put some static stuff in setup.cfg. there was rarely a rigorous reason about what went in one but not the other

SyntaxColoring commented on hardware/pyproject.toml:
```diff
@@ -0,0 +1,73 @@
+root = '..'
+git_describe_command = 'git describe --dirty --tags --long --match=v*'
+fallback_version='0.0.0-dev'
+# Note: these are required for setuptools_scm to not automatically bump the version used in the package
+# from the latest version it detects. This would actually be kind of useful, but we can only do it if
+# everything is doing it since everything is coversioned (otherwise it breaks pipenv sync)
```
FYI this nice comment is present in some of the pyproject.tomls but not all of them.

SyntaxColoring reviewed (approved):
Yay. :)

vegano1 reviewed (approved):
This looks good to me
- Locally building works from a teardown/setup
- Locally building works from a fresh clone
- I was able to push python to Flex/OT2
- I was able to build oe-core locally

sfoster1 committed:
fix: limit zipp to 3.21.0

zipp 3.22.0 introduces a build dependency called coherent.licensed,
which per its docs is not safe to have generally installed and should
only be used in an isolated build environment - it will modify any
ongoing build process, for some reason. we don't do isolated build
environments in our system builders, so limit zipp to 3.21.0

sfoster1 committed:
fix: do not need legacy editable

And it now messes with buildroot

sfoster1 committed:
fix: a couple more dependency tweaks

sfoster1 committed:
fix: copy our nice comments around

sfoster1 commented:
Did this via https://github.com/Opentrons/opentrons/pull/18538
